File: 0113-path-sum-ii/0113-path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        if not root:
            return []
        res = []

        def dfs(node, sm, lst):
            if not node:
                return []
        
            lst.append(node.val)
            if not node.left and not node.right and sm == targetSum:
                res.append(list(lst))
            else:
                if node.left:
                    dfs(node.left, sm+node.left.val, lst)
                if node.right:
                    dfs(node.right, sm+node.right.val, lst)
            lst.pop()
        
        dfs(root, root.val, [])
        return res
        

File: 0113-path-sum-ii/0113-path-sum-ii.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function pathSum(root: TreeNode | null, targetSum: number): number[][] {
    if(!root) return [];
    let res: number[][] = [];
    
    let dfs = (node: TreeNode, sm: number, lst: number[]) => {
        if(!node) return;
        lst.push(node.val);

        if(!node.left && !node.right && sm === targetSum) {
            res.push([...lst]);
        } else {
            dfs(node.left, sm+(node.left?.val ?? 0), lst)
            dfs(node.right, sm+(node.right?.val ?? 0), lst)
        }

        lst.pop();
    }

    dfs(root, root.val, []);
    return res;
}; 